### PR TITLE
[Trie] trie scanner

### PR DESCRIPTION
**Pull request by adrian-wang**
## What changes were proposed in this pull request?

add scanner to traverse the trie structure.

## How was this patch tested?

added tests in `PermutermScannerSuite`

### CODE REVIEW

**Comment by adrian-wang** [comment]
I'll merge this by today, so please review it early...

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/PermutermScanner.scala [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    indexData = FiberCacheManager.getOrElseUpdate(indexFiber, conf)
+    val root = open(indexData.buffer, keySchema, indexFile.version(conf))
+
+    _init(root)
+    this
+  }
+
+  def _init(root: TrieNode): Unit = traverse(root, 0)
+  def traverse(trieNode: TrieNode, matching: Int): Unit = {
+    if (pattern.length == matching) {
```
Comment (yueawang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/PermutermScanner.scala]: Is it possible that pattern null or empty? null may raises exception, and empty returns the whole tree.

Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/PermutermScanner.scala]: `pattern` will be fed by index selection part, and have itself secured there

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/PermutermScanner.scala [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    val dataEnd = Platform.getInt(baseObject, baseOffset + data.size - 8)
+    val length = Platform.getInt(baseObject, baseOffset + data.size - 4)
+    val rootPage = Platform.getInt(baseObject, baseOffset + data.size - 12)
+    val rootOffset = Platform.getInt(baseObject, baseOffset + data.size - 16)
+    UnsafeTrie(data, rootPage, rootOffset, dataEnd)
+  }
+
+  override def toString: String = "PermutermScanner"
+
+  override def hasNext: Boolean = if (matchRoot != null) {
```
[the remark this answers is not in the document]
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/PermutermScanner.scala]: Yes, that would be better

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/PermutermScanner.scala [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        case None => matchRoot = null
+        case Some(node) => traverse(node, matching + 1)
+      }
+    }
+  }
+
+  def open(
+    data: ChunkedByteBuffer,
+    keySchema: StructType,
+    version: Int = IndexFile.INDEX_VERSION): TrieNode = {
```
[the remark this answers is not in the document]
Comment (adrian-wang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/PermutermScanner.scala]: Let's clean up this when conforming interface between different index types.

**Pull request merged**